### PR TITLE
feat: add begin and end blocker support to module/server

### PR DESCRIPTION
**Pull request by robert-zaremba**
## Description

Closes: #689

Initially I started implementing begin and end blocker support through new `Configurator` register functions (similarly to `RegisterGenesisHandlers). However this  approach has downsides:
* it hides modules, so it's not possible to customize the execution order
* it further away from the SDK design
* would require yet another small rework of the ecocredit Module by removing the keeper and moving the whole logic to the keeper itself (the server) - which is mixing module logic with the data and execution layer.

So, instead I added Begin and End blocker functions to the Regen `module/server.Module`, which uses modules directly to execute the begin / end blockers. 
I had to update the module to be mutable - the keeper assignment in [RegisterServices](https://github.com/regen-network/regen-ledger/blob/26cb432c298245804dd56360437f7f96977cd39b/x/ecocredit/module/module.go#L65) wouldn't  work otherwise.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/ecocredit/module/module.go [hunk 1]
```diff
@@ -35,23 +35,23 @@ type Module struct {
 }
 
 // NewModule returns a new Module object.
-func NewModule(paramSpace paramtypes.Subspace, accountKeeper ecocredit.AccountKeeper, bankKeeper ecocredit.BankKeeper) Module {
+func NewModule(paramSpace paramtypes.Subspace, accountKeeper ecocredit.AccountKeeper, bankKeeper ecocredit.BankKeeper) *Module {
```
Comment (robert-zaremba) [inline x/ecocredit/module/module.go:38]: Pointer receiver is required to handle keeper registration in `RegisterServices`

**Review by aaronc — COMMENTED** [review]
It doesn't seem like this does any particular ordering, other than registration order which is deterministic. Maybe that's okay since it's a short term fix. It would be good to actually make ecocredit have a begin and end blocker even if they don't do anything yet.

**Comment by robert-zaremba** [comment]
@aaronc x/ecocredit has a registered Begin Blocker (in this PR). It doesn't have end blocker. I can add an empty one.

>  Maybe that's okay since it's a short term fix. 

I don't think we need it (there is only one module, and even with cosmowasm, registration order will work good enough). Once we will Cosmos SDK 0.46 we should migrate to `AppModule`

**Comment by aaronc** [comment]
the simulation issue does seem to be related to the ecocredit begin blocker

File: x/ecocredit/server/expiration.go [hunk 2]
```diff
@@ -26,16 +28,19 @@ func (s serverImpl) PruneOrders(ctx sdk.Context) error {
...
 		err = s.sellOrderTable.Delete(ctx, sellOrder.OrderId)
 		if err != nil {
+			logger.Error("problem with sell order delete", "err", err)
 			return err
 		}
 	}
 
 	buyOrdersIter, err := s.buyOrderByExpirationIndex.PrefixScan(ctx, minTime, blockTime)
 	if err != nil {
+		logger.Error("problem with buy order iteratorr", "err", err)
```
Comment (technicallyty) [inline x/ecocredit/server/expiration.go:43]:
```suggestion
		logger.Error("problem with buy order iterator", "err", err)
```

**Review by technicallyty — COMMENTED** [review]
could you add a test for these changes?

**Comment by aleem1314** [comment]
@robert-zaremba I think simulations are failing because of duplicate prefixes for orm [tables](https://github.com/regen-network/regen-ledger/blob/master/x/ecocredit/server/server.go#L22-L32).

**Review by ryanchristo — APPROVED** [review]
Looks good to me. One question. Pre-approving.

File: x/ecocredit/server/server.go [hunk 3]
```diff
@@ -22,23 +22,23 @@ const (
...
 	BatchInfoTablePrefix     byte = 0x6
 
-	ProjectInfoTablePrefix    byte = 0x12
-	ProjectInfoTableSeqPrefix byte = 0x13
-	ProjectsByClassIDIndex    byte = 0x14
-	BatchesByProjectIndex     byte = 0x15
+	ProjectInfoTablePrefix    byte = 0x10
+	ProjectInfoTableSeqPrefix byte = 0x11
+	ProjectsByClassIDIndex    byte = 0x12
+	BatchesByProjectIndex     byte = 0x13
```
Comment (aaronc) [inline x/ecocredit/server/server.go:28]: I don't think this file should have changed in this PR. Looks like something got messed up

**Review by aaronc — CHANGES_REQUESTED** [review]
LGTM. But there are some unrelated changes that need to be reverted. Once that's done ready to approve

File: x/ecocredit/server/server.go (hunk 3 above)
Comment (robert-zaremba) [reply x/ecocredit/server/server.go:28]: This prevents the sim tests to pass.

Comment (robert-zaremba) [reply x/ecocredit/server/server.go:28]: basically there is a conflict between the prefixes.

Comment (robert-zaremba) [reply x/ecocredit/server/server.go:28]: this got broken in https://github.com/regen-network/regen-ledger/pull/647

**Review by aaronc — APPROVED** [review]
approving as @robert-zaremba  clarified that the changes to store key prefixes are to fix conflicts introduced in #647

**Pull request merged**